### ISSUE TITLE
Allow easy consumption of feature branches

### ISSUE THREAD

**Issue by kzu** [dotnet/roslyn#45183]
Currently, [consuming VSIX and nuget packages](https://github.com/dotnet/roslyn/blob/master/docs/contributing/Building%2C%20Debugging%2C%20and%20Testing%20on%20Windows.md#deploying-with-vsix-and-nuget-package) is mostly focused on locally built bits. For CI builds, we only offer the pipeline artifacts. Staying up to date with a quickly evolving feature (such as [records](https://github.com/dotnet/roslyn/tree/features/records)) is a non-trivial commitment :).

The proposal is to provide automatic VSIX gallery feeds for feature branches, so that staying up-to-date is trivial (just let VS offer/install the updates), as well as a corresponding NuGet feed with proper versioning that matches the produced VSIXes, so you can easily keep both in sync.

Intended experience is:

1. Users download a feature branch VSIX from a permalink URL linked from the readme, simply like https://roslynfeed.azurewebsites.net/latest. The service resolving that URL would use the `Referer` header to automatically redirect to the latest build for the branch the user is currently navigating.

2. Once installed, the VSIX automatically configures a VSIX gallery feed to get updates from:

![image](https://user-images.githubusercontent.com/169707/84607135-f30dbf00-ae81-11ea-9025-ea6ef7cfa77d.png)

3. The auto-configured feed clearly mentions the feature branch (`records` in this case) it belongs to. Clicking on the `More Information` link in the extension description would take you to the source commit for that build. Also, the extension contains a unique version number starting with `42.42.42` (to denote a CI non-production build) and a unique (always incrementing) suffix number.

4. When updates are available, they are easily installed from the usual Extension Manager UI:

![image](https://user-images.githubusercontent.com/169707/84607732-59481100-ae85-11ea-9ac9-1dc0c007822b.png)

5. There is a corresponding NuGet feed (i.e. https://roslyn.azurewebsites.net/nuget/index.json) that contains properly labeled packages according to the feature branch that built them, such as: `Microsoft.Net.Compilers.Toolset.3.7.0-ci.records.1234`. The format would be `-ci.FEATURE.BUILD`, where `FEATURE` matches the name of a feature after the `features/` prefix currently in use, and `BUILD` would match the suffix number in the VSIX too (to allow a precise match as needed).

6. Since the packages are properly labeled using semver2, it would be possible to wildcard the last segment to get always the latest version:

```xml
<PackageReference Include="Microsoft.Net.Compilers.Toolset" Version="3.7.0-ci.records.*" />
```

7. Alternatively, all this can be done manually by:
    * Creating permalink urls with `https://roslynfeed.azurewebsites.net/latest?FEATURE`
    * Setting up a custom gallery feed with `https://roslyn.azurewebsites.net/vsix/FEATURE/atom.xml`

Obviously all this would need to be fully automated and available without any intervention for any branch under 
`feature/*`, `demo/*` and probably also `master`.

## Preview

I have set it all up [according to the readme](https://github.com/kzu/roslyn/blob/features/feeds/src/Tools/FeatureFeed/Readme.md) 
so it can be seen in action at:

1. CI build of packages and VSIX: https://build.azdo.io/kzu/oss/39
2. CI build and deploy of Azure function: https://build.azdo.io/kzu/oss/40
3. Permalink to latest build for feature branch available in the mentioned readme, as well as https://roslynfeed.azurewebsites.net/latest?feeds (explicit querystring)
4. VSIX gallery (Atom) feed at https://roslyn.blob.core.windows.net/vsix/feeds/atom.xml (generated from upload on step 1, by function on step 2)
5. Serverless NuGet feed from step 1 at https://roslyn.blob.core.windows.net/nuget/index.json

**Comment by kzu** [comment]
Working on the PR(s) to implement this :)

**Comment by kzu** [comment]
I can help set up the various resources, just ping me :). Right now it's only building on my DevOps for my fork branches. /cc @jaredpar @tmat

**Comment by jmarolf** [comment]
So the reason I didn't do this in the past was that there were bugs in the VSIX gallery that caused things to not upgrade nested VSIXes correctly (though they will update correctly on build). @madskristensen do nested vsixes correctly auto-update now?

**Comment by jaredpar** [comment]
@jmarolf is correct. We've done essentially this several times in the past but always had to undo it because it leads to corrupted Visual Studio installations. We spent quite a bit of time attempting to track this down but getting reliable repro's of the problem are difficult and hence it impeded the ability of the platform team to make progress here. 

The root of the problem appears to be around our use of VSIX containers for other VSIX installations. The underlying installation are not ordered hence we can get into a state where we have incompatible VSIX installed and the only choice is to essentially repair the installation.

**Comment by madskristensen** [comment]
@jmarolf @jaredpar Right, VS will always update to the latest version regardless of which gallery hosts the extension. If it exist in both the Marketplace and a private gallery, then you'll bounce back and forth to whichever ever gallery provides the highest version of the extension. The fix is to use different VSIX Product IDs so there's only a single of the same extension per gallery/marketplace.

**Comment by jaredpar** [comment]
Think it's more than that. Been about 1-2 years since we did a deep dive into this but part of the problem seemed to be ordering. Mostly that container VSIX (a VSIX which just contains other VSIX) doesn't have a guaranteed order for the nested VSIX. That creates issues with Roslyn because our nested VSIX are order dependent. Sometimes the installer did it in the right order and sometimes it didn't. 

That at least wast the theory based on the outcomes both teams were looking at. We never got a solid repro though hence never made progress on the problem. We haven't shipped to customers though because every time we enabled this it was just a week or so until we ended up with a corrupted Visual Studio.

**Comment by madskristensen** [comment]
@jaredpar Got it, now I understand the issue. I remember something about this coming up back then. Let's ask @mjcavallari

**Comment by kzu** [comment]
@wade0016 do you think an alpha pack might be better for this scenario?

**Comment by tmat** [comment]
Our builds produce `Roslyn_Preview.zip` that contains `install.bat` script that reliably installs/uninstalls all the VSIXes to a specified VS instance. Isn't that sufficient? Any time you want to get the latest you download this file, unzip and run the script. You can probably write a PowerShell scripts that makes it even easier.

**Comment by jmarolf** [comment]
right, I wrote that. The intent is that you:
1. run the script
2. test out the changes
3. uninstall it

having it be a "dogfood channel" isn't really supported. You cannot upgrade from build-to-build

**Comment by jmarolf** [comment]
> alpha pack might be better for this scenario?

perhaps, at the time I looked into this alpha packs where something we were told not to do because they were too new. Stuff has changed since then so it could be time to revisit this. If the platform team is ok with people consuming us as an alpha pack I am all for it.

**Comment by kzu** [comment]
Yes, I saw the script, tried it. It's far from the typical experience you'd get from a VS gallery, though :). Just finding the right artifact takes multiple clicks, downloading, unzipping (twice, once for the VSIX another for the packages), running the scripts... it's not really convenient. It's also not easy to know when/what is new either, so, not ideal. 

The fact that the VSIX has been broken for two months already (since #43208, the manifest points to a non-existent embedded VSIX so it can't be installed at all) and nobody noticed, is probably a good indication that as a solution, it's not very discoverable or usable. 

Also, it's not entirely clear to me why you need a custom `VsixExpInstaller` tool. If the "double click" experience for a VSIX is broken, maybe we could invest in fixing it. 

I think it's worth having something as built-in VS as possible and as streamlined as possible to get early feedback and excitement around upcoming features, with very low friction to trying them out.

**Comment by jmarolf** [comment]
I agree with all your points, unfortunately the last time I looked at this I could not find a way forward. I know the VS extensibility team is rather busy and may not have time to fix the issues that force us to use our own version of the `VsixExpInstaller`.

**Comment by jaredpar** [comment]
@jmarolf summed up my thoughts. Part of the problem though is we're still not 100% sure what the issue is. Some period after we enable auto-updates, a day or a week, the instance gets corrupted and we have to repair to move forward. 

We tried a number of different debugging approaches but couldn't find a solid repro for the extensibility team to make progress on.  If there was a stable repro we could give them it's possible this could help move the fix forward.

Another aspect of our old decisions here to keep in mind is that even if the bug is fixed there are still two other problems to consider:

1. VS is a moving target and Roslyn often has to make changes to remain functional on VS. Our feature branches are not guaranteed, or required, to be up to date with the main branch. That means it's not uncommon for feature branches to be unusable on current VS due to these changes. To make this usable we'd have to basically commit to a stronger up to date discipline in those branches + limit their VSIX publishing to a specific feed of VS (RTM, Preview, etc ...).
1. Roslyn is an API that other teams build on. It's possible for instance that subtle changes in Roslyn will end up breaking F# or Typescript. When that happens it's typically quickly fixed and customers never really see that. If we have essentially nightly VSIX adoption though it's possible customers will see this. 

All of these issues are part of the reason we introduced the `/preview` flag. This lets us move features into the main VS tree much sooner than we previously would. Basically once a feature is approaching done we move it into master and begin getting feedback. At times this lets us release features nearly a year before they actually RTM. 

This acquisition story avoids all the issues we hit trying to get feedback through VSIX. To preview a feature just update your `<LangVersion>` and go. This certainly isn't as rapid is "update the day after a commit into a feature branch" but it's helped us in getting feedback from customers sooner.
